node/actors/bft/src/replica/tests.rs

```diff
@@ -1,54 +1,477 @@
+//! - [x] leader_prepare_proposal_oversized_payload
+//! - [x] leader_prepare_proposal_mismatched_payload
+//! - [x] leader_prepare_proposal_when_previous_not_finalized
+//! - [x] leader_prepare_proposal_invalid_parent_hash
+//! - [x] leader_prepare_proposal_non_sequential_number
+//! - [x] leader_prepare_reproposal_without_quorum
+//! - [x] leader_prepare_reproposal_when_finalized
+//! - [x] leader_prepare_reproposal_invalid_block
+//!
+//! - [x] leader_commit_sanity
+//! - [x] leader_commit_sanity_yield_replica_prepare
+//! - [x] leader_commit_invalid_leader
+//! - [x] leader_commit_old
+//! - [x] leader_commit_invalid_sig
+//! - [x] leader_commit_invalid_commit_qc
+//!
+use super::{
+    leader_commit::Error as LeaderCommitError, leader_prepare::Error as LeaderPrepareError,
+};
+use crate::{
+    inner::ConsensusInner,
+    leader::ReplicaPrepareError,
+    misc::{consensus_threshold, faulty_replicas},
+    testonly::ut_harness::UTHarness,
+};
+use assert_matches::assert_matches;
 use rand::Rng;
-use zksync_concurrency::{ctx, scope, testonly::abort_on_panic, time};
-use zksync_consensus_network::io::{ConsensusInputMessage, Target};
-use zksync_consensus_roles::validator::{self, ViewNumber};
-
-#[tokio::test]
-async fn start_new_view_not_leader() {
-    abort_on_panic();
-    let ctx = &ctx::test_root(&ctx::ManualClock::new());
-    let rng = &mut ctx.rng();
-
-    let keys: Vec<_> = (0..4).map(|_| rng.gen()).collect();
-    let (genesis, val_set) = testonly::make_genesis(&keys, validator::Payload(vec![]));
-    let (mut consensus, mut pipe) =
-        testonly::make_consensus(ctx, &keys[0], &val_set, &genesis).await;
-    // TODO: this test assumes a specific implementation of the leader schedule.
-    // Make it leader-schedule agnostic (use epoch to select a specific view).
-    consensus.replica.view = ViewNumber(1);
-    consensus.replica.high_qc = rng.gen();
-    consensus.replica.high_qc.message.view = ViewNumber(0);
-
-    scope::run!(ctx, |ctx, s| {
-        s.spawn(async {
-            consensus
-                .replica
-                .start_new_view(ctx, &consensus.inner)
-                .await
-                .unwrap();
-            Ok(())
+use std::cell::RefCell;
+use zksync_consensus_roles::validator::{
+    BlockHeaderHash, ConsensusMsg, LeaderPrepare, Payload, PrepareQC, ReplicaCommit,
+    ReplicaPrepare, ViewNumber,
+};
+
+#[tokio::test]
+async fn leader_prepare_sanity() {
+    let mut util = UTHarness::new().await;
+
+    let leader_prepare = util.new_procedural_leader_prepare().await;
+    util.dispatch_leader_prepare(leader_prepare).await.unwrap();
+}
+
+#[tokio::test]
+async fn leader_prepare_sanity_yield_replica_commit() {
+    let mut util = UTHarness::new().await;
+
+    let leader_prepare = util.new_procedural_leader_prepare().await;
+    util.dispatch_leader_prepare(leader_prepare).await.unwrap();
+    util.recv_signed()
+        .await
+        .unwrap()
+        .cast::<ReplicaCommit>()
+        .unwrap();
+}
```

Comment:
You want to add sanity tests for reproposals too. Now we are just testing new proposals.

Reply:
Added here: https://github.com/matter-labs/era-consensus/pull/39/commits/d02ee90607d30b45cec52d11d5a007be2e23d7e0

* `replica_prepare_sanity_yield_leader_prepare_reproposal` is focused on asserting the produced `leader_prepare` msg fields.
* `leader_prepare_reproposal_sanity` only checks that `replica_prepare` processing produced a reproposal, and then verifies a successful `leader_prepare` processing of it by the replica. 

I did not add a test for a reproposal with insufficiant quorum because it is already covered by `leader_prepare_reproposal_without_quorum`.